File: source/downloads/code/LeetCode/Python/Mini-Parser.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger(object):
#    def __init__(self, value=None):
#        """
#        If value is not specified, initializes an empty list.
#        Otherwise initializes a single integer equal to value.
#        """
#
#    def isInteger(self):
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        :rtype bool
#        """
#
#    def add(self, elem):
#        """
#        Set this NestedInteger to hold a nested list and adds a nested integer elem to it.
#        :rtype void
#        """
#
#    def setInteger(self, value):
#        """
#        Set this NestedInteger to hold a single integer equal to value.
#        :rtype void
#        """
#
#    def getInteger(self):
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        :rtype int
#        """
#
#    def getList(self):
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        :rtype List[NestedInteger]
#        """



class Solution(object):
    def deserialize(self, s):
        """
        :type s: str
        :rtype: NestedInteger
        """
        # 数字
        nums = None
        # 正负
        neg = 1

        stack = []
        # 只含有数字
        if s[0] != '[':
            return NestedInteger(int(s))


        for c in s:
            if c == '-':
                neg = -1
            elif c in '0123456789':
                if nums == None:
                    nums = int(c)
                else:
                    nums = 10*nums + int(c)
            elif c == '[':
                stack.append(NestedInteger())
            else:
            # "," or "]"
                if nums != None:
                    stack[-1].add(NestedInteger(nums*neg))
                    nums = None
                    neg = 1
                if c == ']':
                    top = stack.pop()
                    if stack:
                        stack[-1].add(top)
                    if not stack:
                        return top

File: source/downloads/code/LeetCode/Java/Mini-Parser.java
public class Solution {
    public NestedInteger deserialize(String s) {
    
        if(s.isEmpty()) return null;
        // 只有一个int string -> int
        if(s.charAt(0) != '[') return new NestedInteger(Integer.valueOf(s));

        Stack<NestedInteger> stack = new Stack<NestedInteger>();
        NestedInteger cur = null;
        int l = 0;

        for(int r=0; r<s.length(); r++){
            char ch = s.charAt(r);
            // 把当前cur压入栈，并开始一个新的
            if(ch == '['){
                if(cur != null)
                    stack.push(cur);
                cur = new NestedInteger();
                l = r+1;
            }
            // 结束当前NestedInteger，弹出一个
            else if (ch==']'){
                String num = s.substring(l,r);
                if(!num.isEmpty()) cur.add(new NestedInteger(Integer.valueOf(num)));

                if(!stack.isEmpty()){
                    NestedInteger top = stack.pop();
                    top.add(cur);
                    cur = top;
                }
                l = r+1;
            }
            // 向cur中插入一个新num
            else if (ch==','){
                // 有数字
                if(s.charAt(r-1)!=']'){
                    String num = s.substring(l,r);
                    cur.add(new NestedInteger(Integer.valueOf(num)));
                }
                l = r+1;
            
            }
        
        }    
        return cur;
    }
}
